src/app/cmd/flatten_layers.cpp

```diff
@@ -97,44 +127,89 @@ void FlattenLayers::onExecute()
 
-      // TODO Keep cel links when possible
+      // Get exact bounds for rendered frame
+      gfx::Rect bounds = image->bounds();
+      const bool shrink = doc::algorithm::shrink_bounds(
+        image.get(), image->maskColor(), nullptr,
+        image->bounds(), bounds);
 
-      ImageRef cel_image;
+      // Skip when fully transparent
       Cel* cel = flatLayer->cel(frame);
+      if (!shrink) {
+        if (!newFlatLayer && cel)
+          executeAndAdd(new cmd::RemoveCel(cel));
+
+        continue;
+      }
+
+      // Apply shrunk bounds to new image
+      const ImageRef new_image(doc::crop_image(
+        image.get(), bounds, image->maskColor()));
+
+      // Replace image on existing cel
       if (cel) {
+
+        // TODO Keep cel links when possible
+
         if (cel->links())
           executeAndAdd(new cmd::UnlinkCel(cel));
 
-        cel_image = cel->imageRef();
+        const ImageRef cel_image = cel->imageRef();
         ASSERT(cel_image);
 
-        executeAndAdd(
-          new cmd::CopyRect(cel_image.get(), image.get(),
-                            gfx::Clip(0, 0, image->bounds())));
+        // Reset cel properties when flattening in-place
+        if (!newFlatLayer) {
+          executeAndAdd(new cmd::SetCelOpacity(cel, 255));
+          executeAndAdd(new cmd::SetCelPosition(cel,
+            area.x+bounds.x, area.y+bounds.y));
+        }
+
+        // Modify destination cel
+        executeAndAdd(new cmd::ReplaceImage(sprite, cel_image, new_image));
       }
+      // Add new cel on null
       else {
-        gfx::Rect bounds(image->bounds());
-        if (doc::algorithm::shrink_bounds(
-              image.get(), image->maskColor(), nullptr, bounds)) {
-          cel_image.reset(
-            doc::crop_image(image.get(), bounds, image->maskColor()));
-          cel = new Cel(frame, cel_image);
-          cel->setPosition(bounds.origin());
+        cel = new Cel(frame, new_image);
+        cel->setPosition(area.x+bounds.x, area.y+bounds.y);
+
+        // No need to undo adding this cel when flattening onto
+        // a new layer, as the layer itself would be destroyed,
+        // hence the lack of a command
+        if (newFlatLayer) {
           flatLayer->addCel(cel);
         }
+        else {
+          executeAndAdd(new cmd::AddCel(flatLayer, cel));
+        }
       }
     }
   }
 
+  // Notify observers when merging down
+  if (m_options.mergeDown)
+    doc->notifyLayerMergedDown(list.back(), flatLayer);
+
   // Add new flatten layer
-  if (newFlatLayer)
-    executeAndAdd(new cmd::AddLayer(list.front()->parent(), flatLayer, list.front()));
+  if (newFlatLayer) {
+    executeAndAdd(new cmd::AddLayer(
+      list.front()->parent(), flatLayer, list.front()));
+
+  }
+  // Reset layer properties when flattening in-place
+  else {
+    executeAndAdd(new cmd::SetLayerOpacity(flatLayer, 255));
+    executeAndAdd(new cmd::SetLayerBlendMode(
+      flatLayer, doc::BlendMode::NORMAL));
```

Comment:
Same here as with the cel opacity:
```suggestion
    if (flatLayer->opacity() != 255)
      executeAndAdd(new cmd::SetLayerOpacity(flatLayer, 255));

    if (flatLayer->blendMode() != doc::BlendMode::NORMAL) {
      executeAndAdd(new cmd::SetLayerBlendMode(
                      flatLayer, doc::BlendMode::NORMAL));
    }
```